File: internal/agent/tmux.go
package agent

import (
	"fmt"
	"os/exec"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/leo/agent-mux/internal/provider"
)

// rawPane holds parsed tmux pane info before status detection.
type rawPane struct {
	target, session, window, windowName, pane, path, cmd string
	pid                                                  int
	windowActivity                                       int64
}

// parseTmuxPanes parses tmux list-panes output into rawPane structs.
func parseTmuxPanes(out []byte) []rawPane {
	var raw []rawPane
	for line := range strings.SplitSeq(strings.TrimSpace(string(out)), "\n") {
		if line == "" {
			continue
		}
		fields := strings.SplitN(line, "\t", 6)
		if len(fields) < 6 {
			continue
		}
		target, cmd, path, pidStr, windowName, actStr := fields[0], fields[1], fields[2], fields[3], fields[4], fields[5]
		pid, _ := strconv.Atoi(pidStr)
		activity, _ := strconv.ParseInt(actStr, 10, 64)
		session, window, pane := ParseTarget(target)
		raw = append(raw, rawPane{target, session, window, windowName, pane, path, cmd, pid, activity})
	}
	return raw
}

// resolveAgentPanes filters raw panes to only those running a registered agent.
// Uses the process table to resolve agents that run under a generic command
// (e.g. gemini runs as "node").
func resolveAgentPanes(raw []rawPane, pt *provider.ProcessTable) []rawPane {
	var agents []rawPane
	for _, r := range raw {
		cmd := provider.Resolve(r.cmd, r.pid, pt)
		if cmd == "" {
			continue
		}
		r.cmd = cmd
		agents = append(agents, r)
	}
	return agents
}

// listTmuxPanes runs tmux list-panes and returns raw output.
func listTmuxPanes() ([]byte, error) {
	return exec.Command("tmux", "list-panes", "-a", "-F",
		"#{session_name}:#{window_index}.#{pane_index}\t#{pane_current_command}\t#{pane_current_path}\t#{pane_pid}\t#{window_name}\t#{window_activity}").Output()
}

// loadProcessTable snapshots the process tree via a single ps call.
func loadProcessTable() provider.ProcessTable {
	out, err := exec.Command("ps", "-eo", "pid,ppid,comm,args").Output()
	if err != nil {
		return provider.ProcessTable{
			Children: make(map[int][]int),
			Comm:     make(map[int]string),
			Args:     make(map[int]string),
		}
	}
	return provider.ParseProcessTable(string(out))
}

// ListPanesBasic returns panes with StatusIdle (no status detection).
// Used for instant initial display before async status detection kicks in.
func ListPanesBasic() ([]Pane, error) {
	var (
		tmuxOut []byte
		tmuxErr error
		history map[string]time.Time
		pt      provider.ProcessTable
	)
	var wg sync.WaitGroup
	wg.Add(3)
	go func() {
		defer wg.Done()
		tmuxOut, tmuxErr = listTmuxPanes()
	}()
	go func() {
		defer wg.Done()
		history = LastActiveByProject()
	}()
	go func() {
		defer wg.Done()
		pt = loadProcessTable()
	}()
	wg.Wait()

	if tmuxErr != nil {
		return nil, fmt.Errorf("tmux list-panes: %w", tmuxErr)
	}

	raw := resolveAgentPanes(parseTmuxPanes(tmuxOut), &pt)
	panes := make([]Pane, len(raw))
	for i, r := range raw {
		panes[i] = Pane{
			Target:         r.target,
			Session:        r.session,
			Window:         r.window,
			WindowName:     r.windowName,
			Pane:           r.pane,
			Path:           r.path,
			PID:            r.pid,
			Status:         StatusIdle,
			WindowActivity: r.windowActivity,
			LastActive:     history[r.path],
		}
	}
	return panes, nil
}

// ListPanes returns all tmux panes running a registered agent with full
// status detection. Runs tmux list-panes, history read, and process table
// snapshot in parallel, then checks attention heuristics per pane.
func ListPanes() ([]Pane, error) {
	var (
		tmuxOut []byte
		tmuxErr error
		history map[string]time.Time
		pt      provider.ProcessTable
	)
	var wg sync.WaitGroup
	wg.Add(3)
	go func() {
		defer wg.Done()
		tmuxOut, tmuxErr = listTmuxPanes()
	}()
	go func() {
		defer wg.Done()
		history = LastActiveByProject()
	}()
	go func() {
		defer wg.Done()
		pt = loadProcessTable()
	}()
	wg.Wait()

	if tmuxErr != nil {
		return nil, fmt.Errorf("tmux list-panes: %w", tmuxErr)
	}

	raw := resolveAgentPanes(parseTmuxPanes(tmuxOut), &pt)

	panes := make([]Pane, len(raw))
	for i, r := range raw {
		panes[i] = Pane{
			Target:         r.target,
			Session:        r.session,
			Window:         r.window,
			WindowName:     r.windowName,
			Pane:           r.pane,
			Path:           r.path,
			PID:            r.pid,
			Status:         StatusIdle,
			WindowActivity: r.windowActivity,
			LastActive:     history[r.path],
		}
	}

	// Run attention heuristics and git enrichment concurrently.
	var allWg sync.WaitGroup
	allWg.Go(func() {
		EnrichPanes(panes)
	})
	for i := range panes {
		allWg.Go(func() {
			panes[i].HeuristicAttention = checkAttention(panes[i].Target)
		})
	}
	allWg.Wait()
	return panes, nil
}

// checkAttention captures the last 10 lines of a pane and returns whether
// the content matches attention heuristics (waiting for user input).
func checkAttention(target string) bool {
	return needsAttention(capturePaneLines(target))
}

// needsAttention checks if a pane is waiting for user interaction.
func needsAttention(lines []string) bool {
	content := strings.Join(lines, "\n")
	for _, pattern := range []string{
		"Do you want to proceed?",
		"Do you want to allow",
		"Allow once",
		"press Enter to approve",
		"Enter to select",
		"Type something",
		"Esc to cancel",
		"I'll wait for your",
		"waiting for your response",
		"Let me know when",
		"Please let me know",
		"What would you like",
		"How would you like",
		"Should I proceed",
		"Would you like me to",
		"please provide",
		"please specify",
		"I need more information",
		"Could you clarify",
		"awaiting your",
		"ready when you are",
		"let me know if you'd like",
		"Feel free to ask",
		"Is there anything else",
		"What else can I help",
		"Want me to",
		"Shall I",
		"Do you want me to",
		"Ready to proceed",
	} {
		if strings.Contains(content, pattern) {
			return true
		}
	}
	for i := len(lines) - 1; i >= 0; i-- {
		line := strings.TrimSpace(lines[i])
		if line == "" {
			continue
		}
		if strings.HasSuffix(line, "?") && !strings.HasPrefix(line, "❯") {
			return true
		}
	}
	return false
}

// capturePaneLines captures the last 10 visible lines of a tmux pane.
func capturePaneLines(target string) []string {
	out, err := exec.Command("tmux", "capture-pane", "-t", target, "-p", "-S", "-10").Output()
	if err != nil {
		return nil
	}
	return strings.Split(strings.TrimRight(string(out), "\n"), "\n")
}

// CapturePane captures the visible content of a tmux pane.
func CapturePane(target string, lines int) (string, error) {
	out, err := exec.Command("tmux", "capture-pane", "-t", target, "-e", "-p", "-S",
		fmt.Sprintf("-%d", lines)).Output()
	if err != nil {
		return "", fmt.Errorf("capture-pane %s: %w", target, err)
	}
	return string(out), nil
}

// SwitchToPane switches the tmux client to the given pane.
func SwitchToPane(target string) error {
	session, window, _ := ParseTarget(target)
	sessionWindow := session + ":" + window
	if err := exec.Command("tmux", "switch-client", "-t", sessionWindow).Run(); err != nil {
		return fmt.Errorf("switch-client: %w", err)
	}
	if err := exec.Command("tmux", "select-pane", "-t", target).Run(); err != nil {
		return fmt.Errorf("select-pane: %w", err)
	}
	return nil
}

// KillPane kills a tmux pane. If it's the only pane in the window, kills the window instead.
func KillPane(target string) error {
	session, window, _ := ParseTarget(target)
	sessionWindow := session + ":" + window

	out, err := exec.Command("tmux", "list-panes", "-t", sessionWindow).Output()
	if err != nil {
		return fmt.Errorf("list-panes: %w", err)
	}
	paneCount := len(strings.Split(strings.TrimSpace(string(out)), "\n"))

	if paneCount <= 1 {
		return exec.Command("tmux", "kill-window", "-t", sessionWindow).Run()
	}
	return exec.Command("tmux", "kill-pane", "-t", target).Run()
}

// parseTarget splits "foo:2.1" into session="foo", window="2", pane="1".
func ParseTarget(s string) (session, window, pane string) {
	colonIdx := strings.LastIndex(s, ":")
	if colonIdx < 0 {
		return s, "", ""
	}
	session = s[:colonIdx]
	rest := s[colonIdx+1:]
	dotIdx := strings.LastIndex(rest, ".")
	if dotIdx < 0 {
		return session, rest, ""
	}
	return session, rest[:dotIdx], rest[dotIdx+1:]
}

File: src/agent/tmux.rs
use std::fs::OpenOptions;
use std::process::{Command, Stdio};
use std::sync::{Mutex, OnceLock};
use std::thread;
use std::time::{Duration, Instant};

use anyhow::{Context, Result, anyhow};
use regex::Regex;
use sha2::{Digest, Sha256};

use crate::agent::Pane;
use crate::agent::git::enrich_panes;
use crate::agent::provider::{ProcessTable, parse_process_table, resolve};

const PROCESS_TABLE_TTL: Duration = Duration::from_secs(1);

#[derive(Debug, Clone)]
struct ProcessTableCache {
    loaded_at: Instant,
    table: ProcessTable,
}

#[derive(Debug, Clone)]
struct RawPane {
    pane_id: String,
    target: String,
    session: String,
    window: String,
    window_name: String,
    pane: String,
    path: String,
    cmd: String,
    pid: i32,
    window_focused: bool,
}

pub fn list_panes() -> Result<Vec<Pane>> {
    let _g = smelt_perf::perf::begin("agent.list_panes");
    let mut panes = list_panes_fast()?;
    enrich_panes(&mut panes);
    Ok(panes)
}

pub fn list_panes_fast() -> Result<Vec<Pane>> {
    let _g = smelt_perf::perf::begin("agent.list_panes_fast");
    let mut panes = fetch_panes()?;
    capture_content(&mut panes);
    Ok(panes)
}

fn fetch_panes() -> Result<Vec<Pane>> {
    let _g = smelt_perf::perf::begin("tmux.fetch_panes");
    let tmux_out = list_tmux_panes()?;
    let pt = load_process_table();
    let raw = {
        let _g = smelt_perf::perf::begin("provider.resolve_panes");
        resolve_agent_panes(parse_tmux_panes(&tmux_out), &pt)
    };
    smelt_perf::perf::record_value("tmux.agent_panes", raw.len() as u64);
    Ok(raw
        .into_iter()
        .enumerate()
        .map(|(order, r)| Pane {
            pane_id: r.pane_id,
            target: r.target,
            session: r.session,
            window: r.window,
            window_name: r.window_name,
            pane: r.pane,
            path: r.path,
            pid: r.pid,
            window_active: r.window_focused,
            order,
            provider: r.cmd,
            ..Pane::default()
        })
        .collect())
}

fn list_tmux_panes() -> Result<String> {
    let _g = smelt_perf::perf::begin("tmux.list_panes");
    let out = Command::new("tmux")
        .arg("list-panes")
        .arg("-a")
        .arg("-F")
        .arg("#{session_name}:#{window_index}.#{pane_index}\t#{pane_current_command}\t#{pane_current_path}\t#{pane_pid}\t#{window_name}\t#{window_active}#{?session_attached,1,0}#{pane_active}\t#{pane_id}")
        .output()
        .context("tmux list-panes")?;
    if !out.status.success() {
        return Err(anyhow!("tmux list-panes exited with {}", out.status));
    }
    Ok(String::from_utf8_lossy(&out.stdout).into_owned())
}

fn parse_tmux_panes(out: &str) -> Vec<RawPane> {
    let panes: Vec<RawPane> = out
        .trim()
        .lines()
        .filter_map(|line| {
            if line.is_empty() {
                return None;
            }
            let fields: Vec<&str> = line.splitn(7, '\t').collect();
            if fields.len() < 7 {
                return None;
            }
            let (session, window, pane) = parse_target(fields[0]);
            Some(RawPane {
                target: fields[0].to_string(),
                cmd: fields[1].to_string(),
                path: fields[2].to_string(),
                pid: fields[3].parse().unwrap_or(0),
                window_name: fields[4].to_string(),
                window_focused: fields[5] == "111",
                pane_id: fields[6].to_string(),
                session,
                window,
                pane,
            })
        })
        .collect();
    smelt_perf::perf::record_value("tmux.raw_panes", panes.len() as u64);
    panes
}
fn resolve_agent_panes(raw: Vec<RawPane>, pt: &ProcessTable) -> Vec<RawPane> {
    raw.into_iter()
        .filter_map(|mut r| {
            let cmd = resolve(&r.cmd, r.pid, pt);
            if cmd.is_empty() {
                None
            } else {
                r.cmd = cmd;
                Some(r)
            }
        })
        .collect()
}

fn load_process_table() -> ProcessTable {
    static CACHE: OnceLock<Mutex<Option<ProcessTableCache>>> = OnceLock::new();
    let cache = CACHE.get_or_init(|| Mutex::new(None));

    if let Ok(cache) = cache.lock()
        && let Some(entry) = cache.as_ref()
        && entry.loaded_at.elapsed() < PROCESS_TABLE_TTL
    {
        smelt_perf::perf::record_value("process.ps_cache_hit", 1);
        return entry.table.clone();
    }

    let _g = smelt_perf::perf::begin("process.ps");
    let table = Command::new("ps")
        .arg("-eo")
        .arg("pid=,ppid=,command=")
        .output()
        .map(|out| parse_process_table(&String::from_utf8_lossy(&out.stdout)))
        .unwrap_or_default();

    if let Ok(mut cache) = cache.lock() {
        *cache = Some(ProcessTableCache {
            loaded_at: Instant::now(),
            table: table.clone(),
        });
    }
    table
}

fn capture_content(panes: &mut [Pane]) {
    let _g = smelt_perf::perf::begin("tmux.capture_content_all");
    thread::scope(|scope| {
        for pane in panes {
            scope.spawn(move || {
                let (hash, moving, attention) = capture_pane_content(&pane.target);
                pane.content_hash = hash;
                pane.content_moving = moving;
                pane.heuristic_attention = attention;
            });
        }
    });
}

fn capture_pane_content(target: &str) -> (String, bool, bool) {
    let _g = smelt_perf::perf::begin("tmux.capture_pane_content");
    let Ok(out) = Command::new("tmux")
        .arg("capture-pane")
        .arg("-t")
        .arg(target)
        .arg("-p")
        .arg("-S")
        .arg("-10")
        .output()
    else {
        return (String::new(), false, false);
    };
    let content = trim_trailing_newlines(out.stdout);
    smelt_perf::perf::record_value("tmux.capture_bytes", content.len() as u64);
    let hash = short_hash(&content);
    let attention = attention_re().is_match(&String::from_utf8_lossy(&content));
    (hash, false, attention)
}

fn trim_trailing_newlines(mut data: Vec<u8>) -> Vec<u8> {
    while data.last().is_some_and(|b| *b == b'\n') {
        data.pop();
    }
    data
}

fn short_hash(data: &[u8]) -> String {
    let digest = Sha256::digest(data);
    digest[..8].iter().map(|b| format!("{b:02x}")).collect()
}

fn attention_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"Do you want to proceed\?|Do you want to allow|Allow once|press Enter to approve|Enter to select|Type something|Esc to cancel|I'll wait for your|waiting for your response|Let me know when|Please let me know|What would you like|How would you like|Should I proceed|Would you like me to|please provide|please specify|I need more information|Could you clarify|awaiting your|ready when you are|let me know if you'd like|Feel free to ask|Is there anything else|What else can I help|Want me to|Shall I|Do you want me to|Ready to proceed").expect("valid attention regex"))
}

pub fn capture_pane(target: &str, lines: usize) -> Result<String> {
    let _g = smelt_perf::perf::begin("tmux.capture_preview");
    let out = Command::new("tmux")
        .arg("capture-pane")
        .arg("-t")
        .arg(target)
        .arg("-e")
        .arg("-p")
        .arg("-S")
        .arg(format!("-{lines}"))
        .output()
        .with_context(|| format!("capture-pane {target}"))?;
    if !out.status.success() {
        return Err(anyhow!("capture-pane {target} exited with {}", out.status));
    }
    Ok(String::from_utf8_lossy(&out.stdout).into_owned())
}

pub fn switch_to_pane(target: &str) -> Result<()> {
    let (session, window, _) = parse_target(target);
    let session_window = format!("{session}:{window}");
    run_tmux(["switch-client", "-t", &session_window])?;
    run_tmux(["select-pane", "-t", target])
}

pub fn kill_pane(target: &str) -> Result<()> {
    let (session, window, _) = parse_target(target);
    let session_window = format!("{session}:{window}");
    let out = Command::new("tmux")
        .arg("list-panes")
        .arg("-t")
        .arg(&session_window)
        .output()
        .context("list-panes")?;
    let pane_count = String::from_utf8_lossy(&out.stdout).trim().lines().count();
    if pane_count <= 1 {
        run_tmux(["kill-window", "-t", &session_window])
    } else {
        run_tmux(["kill-pane", "-t", target])
    }
}

fn run_tmux<const N: usize>(args: [&str; N]) -> Result<()> {
    let status = Command::new("tmux").args(args).status().context("tmux")?;
    if status.success() {
        Ok(())
    } else {
        Err(anyhow!("tmux exited with {status}"))
    }
}

pub fn start_watch() -> Result<()> {
    if crate::agent::watch::is_running() {
        for _ in 0..5 {
            if crate::agent::ipc::get_state().is_ok_and(|(snapshot, _)| snapshot.is_some()) {
                return Ok(());
            }
            thread::sleep(Duration::from_millis(50));
        }
        stop_watch_process();
        thread::sleep(Duration::from_millis(200));
    }

    std::fs::create_dir_all(crate::agent::persist::state_dir()).context("create state dir")?;
    let log = OpenOptions::new()
        .create(true)
        .append(true)
        .open(crate::agent::watch::log_path())
        .context("open watch log")?;
    let stderr = log.try_clone().context("clone watch log")?;

    let exe = std::env::current_exe().context("current executable")?;
    Command::new(exe)
        .arg("watch")
        .stdin(Stdio::null())
        .stdout(Stdio::from(log))
        .stderr(Stdio::from(stderr))
        .spawn()
        .context("start watch")?;

    for _ in 0..10 {
        if crate::agent::ipc::get_state().is_ok_and(|(snapshot, _)| snapshot.is_some()) {
            return Ok(());
        }
        thread::sleep(Duration::from_millis(50));
    }

    Ok(())
}

pub fn restart_watch() -> Result<()> {
    stop_watch_process();
    thread::sleep(Duration::from_millis(200));
    start_watch()
}

fn stop_watch_process() {
    if let Ok(data) = std::fs::read_to_string(crate::agent::watch::lock_path())
        && let Ok(pid) = data.trim().parse::<i32>()
        && pid > 0
    {
        let _ = Command::new("kill")
            .arg("-TERM")
            .arg(pid.to_string())
            .status();
    }
}

pub fn parse_target(s: &str) -> (String, String, String) {
    let Some(colon_idx) = s.rfind(':') else {
        return (s.to_string(), String::new(), String::new());
    };
    let session = s[..colon_idx].to_string();
    let rest = &s[colon_idx + 1..];
    let Some(dot_idx) = rest.rfind('.') else {
        return (session, rest.to_string(), String::new());
    };
    (
        session,
        rest[..dot_idx].to_string(),
        rest[dot_idx + 1..].to_string(),
    )
}
